control-plane-operator/controllers/awsprivatelink/awsprivatelink_controller.go

```diff
@@ -633,6 +637,137 @@ func (r *AWSEndpointServiceReconciler) reconcileAWSEndpointService(ctx context.C
 	return nil
 }
 
+func (r *AWSEndpointServiceReconciler) reconcileAWSEndpointSecurityGroup(ctx context.Context, awsEndpointService *hyperv1.AWSEndpointService, hcp *hyperv1.HostedControlPlane) error {
+	var sgID string
+	var sg *ec2.SecurityGroup
+
+	log := ctrl.LoggerFrom(ctx)
+	describeSGResult, err := r.ec2Client.DescribeSecurityGroupsWithContext(ctx, &ec2.DescribeSecurityGroupsInput{Filters: vpcEndpointSecurityGroupFilter(hcp.Spec.InfraID, awsEndpointService.Name)})
+	if err != nil {
+		return fmt.Errorf("cannot list security groups, code: %s", supportawsutil.AWSErrorCode(err))
+	}
+	if len(describeSGResult.SecurityGroups) > 0 {
+		sg = describeSGResult.SecurityGroups[0]
+		sgID = aws.StringValue(sg.GroupId)
+		if awsEndpointService.Status.SecurityGroupID != sgID {
+			log.Info("WARNING: found existing security group, but it doesn't match status id, updating", "id", sgID)
+			awsEndpointService.Status.SecurityGroupID = sgID
+		}
+	}
+	if sgID == "" {
+		var err error
+		if sg, err = r.createSecurityGroup(ctx, awsEndpointService, hcp); err != nil {
+			return err
+		}
+		sgID = aws.StringValue(sg.GroupId)
+		awsEndpointService.Status.SecurityGroupID = sgID
+	}
+	machineCIDRs := make([]string, len(hcp.Spec.Networking.MachineNetwork))
+	for i, mNet := range hcp.Spec.Networking.MachineNetwork {
+		machineCIDRs[i] = mNet.CIDR.String()
+	}
+	ingressPermissions := supportawsutil.VPCEndpointSecurityGroupRules(machineCIDRs, vpcEndpointPort(awsEndpointService))
+	missingPermissions := diffPermissions(sg.IpPermissions, ingressPermissions)
+	if len(missingPermissions) > 0 {
+		if _, err = r.ec2Client.AuthorizeSecurityGroupIngressWithContext(ctx, &ec2.AuthorizeSecurityGroupIngressInput{
+			GroupId:       aws.String(sgID),
+			IpPermissions: ingressPermissions,
+		}); err != nil {
+			if supportawsutil.AWSErrorCode(err) != "InvalidPermission.Duplicate" {
+				return fmt.Errorf("failed to set security group ingress rules, code: %s", supportawsutil.AWSErrorCode(err))
```

Comment:
Is this the case when perms are manipulated out of band?

Reply:
The `InvalidPermission.Duplicate` happens if somehow the permission we thought we were adding is already there. It could have been done by the same controller but somehow AWS's API gave us outdated info when we described the SG.

If it's not that, then it could be any error that we should return so we can retry